sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -1177,12 +1272,513 @@ public Write withTypeFn(FieldValueExtractFn typeFn) {
+              "{ \"update\" : %s }%n{ \"doc\" : %s, " + "\"doc_as_upsert\" : true }%n",
+              documentMetadata, document);
+        } else if (spec.getUpsertScript() != null) {
+          return String.format(
+              "{ \"update\" : %s }%n{ \"script\" : {\"source\": \"%s\", "
+                  + "\"params\": %s}, \"upsert\" : %s }%n",
+              documentMetadata, spec.getUpsertScript(), document, document);
+        } else {
+          return String.format("{ \"index\" : %s }%n%s%n", documentMetadata, document);
+        }
+      }
+    }
+
+    private static String lowerCaseOrNull(String input) {
+      return input == null ? null : input.toLowerCase();
+    }
+
+    /**
+     * Extracts the components that comprise the document address from the document using the {@link
+     * Write.FieldValueExtractFn} configured. This allows any or all of the index, type and document
+     * id to be controlled on a per document basis. If none are provided then an empty default of
+     * {@code {}} is returned. Sanitization of the index is performed, automatically lower-casing
+     * the value as required by Elasticsearch.
+     *
+     * @param parsedDocument the json from which the index, type and id may be extracted
+     * @return the document address as JSON or the default
+     * @throws IOException if the document cannot be parsed as JSON
+     */
+    private static String getDocumentMetadata(
+        DocToBulk spec, JsonNode parsedDocument, int backendVersion) throws IOException {
+      DocumentMetadata metadata =
+          new DocumentMetadata(
+              spec.getIndexFn() != null
+                  ? lowerCaseOrNull(spec.getIndexFn().apply(parsedDocument))
+                  : null,
+              spec.getTypeFn() != null ? spec.getTypeFn().apply(parsedDocument) : null,
+              spec.getIdFn() != null ? spec.getIdFn().apply(parsedDocument) : null,
+              (spec.getUsePartialUpdate()
+                      || (spec.getUpsertScript() != null && !spec.getUpsertScript().isEmpty()))
+                  ? DEFAULT_RETRY_ON_CONFLICT
+                  : null,
+              spec.getRoutingFn() != null ? spec.getRoutingFn().apply(parsedDocument) : null,
+              backendVersion,
+              spec.getDocVersionFn() != null ? spec.getDocVersionFn().apply(parsedDocument) : null,
+              spec.getDocVersionType());
+      return OBJECT_MAPPER.writeValueAsString(metadata);
+    }
+
+    /** {@link DoFn} to for the {@link DocToBulk} transform. */
+    @VisibleForTesting
+    static class DocToBulkFn extends DoFn<String, String> {
+      private final DocToBulk spec;
+      private int backendVersion;
+
+      public DocToBulkFn(DocToBulk spec) {
+        this.spec = spec;
+      }
+
+      @Setup
+      public void setup() throws IOException {
+        ConnectionConfiguration connectionConfiguration = spec.getConnectionConfiguration();
+        if (spec.getBackendVersion() == null) {
```

Comment:
Hmm that's a good question, I hadn't quite considered that. What are your thoughts on adding a note in docstring for `withBackendVersion` stating that its value will take precedence and will preclude reading the version from the cluster? Then we would need to trust users to select what fits their need beyond that point.  Thoughts?

Reply:
yes + a warning that mismatch would not work

Reply:
The only qualm I have with verifying the version is that it creates a dependency on a cluster being available. I had hoped to keep the ability (depending on explicitly setting `withBackendVersion`) for `DocToBulk` to be fully free from IO/dependency on an available cluster.

I could do the checking in `ElasticsearchIO.Write`. Thoughts?

Reply:
you can't because the check needs the cluster (as you said) and the DoFn needs to have been staged to the beam cluster in  case there are networking firewalling between were pipeline main() is run and were the ES cluster is. In the past the check was done in WriteFn constructor (which runs in main so in spark/flink client or jobmanager) and it showed network issues in some cases. It was put in WriteFn.setup (which is run in taskmanagers) to avoid that. So you need a DoFn and there is no more in Write.

Reply:
I think I was unclear, when I said "+ a warning on version mismatch", I meant adding a javadoc warning not a check in code otherwise we lose the ability to avoid cluster check and keep a dep with the cluster availability indeed.

Reply:
+ if we offer the ability to have _withBackendVersion()_ to write, it needs to be uniform and offer it to read as well.

Reply:
> please to the check with cluster only if the version is not specified and state in the javadoc of withBackendVersion

Done! https://github.com/apache/beam/pull/14347/commits/54f2fbca29b5eac2b74fbca280886c1c15027090

> plus,
if we offer the ability to have withBackendVersion() to write, it needs to be uniform and offer it to read as well.

Do you mean that we need to `withBackendVersion()` for ElasticsearchIO.Read in this PR? Could we do that in a follow-up instead?